[WIP] Meraki REST module - meraki_rest

kbreit opened the pull request:
##### SUMMARY
New module category for Cisco's Meraki cloud infrastructure. This particular module is a simple wrapper for REST calls with more abstracted modules coming in the future.

##### ISSUE TYPE
 - New Module Pull Request

##### COMPONENT NAME
meraki_rest

##### ANSIBLE VERSION
```
ansible 2.6.0 (meraki/meraki_rest 1ee5fb6f02) last updated 2018/02/25 17:59:12 (GMT -500)
  config file = None
  configured module search path = ['/Users/kbreit/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/kbreit/Documents/Programming/ansible/lib/ansible
  executable location = /Users/kbreit/Documents/Programming/ansible/bin/ansible
  python version = 3.5.4 (default, Feb 25 2018, 14:56:02) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```


##### ADDITIONAL INFORMATION
None

kbreit committed:
Initial commit for Meraki modules. This is a plain module for RESTful communication

kbreit committed:
More changes to support HTTP options.

kbreit committed:
Formatting changes to comply with CI checks

kbreit committed:
Fix condition if first character of path is /

kbreit committed:
More debugging and removed a little bit of debugging code

kbreit committed:
Added support for Meraki API key to be specified via environmental variable

kbreit committed:
Added support for src parameter which will allow for data to be passed directly from a file.

kbreit committed:
Added a lot of debugging information to troubleshoot problems with POST methods

kbreit committed:
Formatting changes per ansibot

kbreit commented:
As of now POST requests don’t work.

kbreit committed:
Add output_level parameters

New parameter (output_level) determines how much content to display for
troubleshooting and debugging purposes.

Akasurde commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,240 @@
+
+    # define the available arguments/parameters that a user can pass to
+    # the module
+    module_args = dict(auth_key=dict(type='str', no_log=True),
```
```
auth_key=dict(type='str', no_log=True)
```
->
```
auth_key=dict(type='str', no_log=True, fallback=(env_fallback, ['MERAKI_KEY']),
```

Akasurde commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,240 @@
+
+    # manipulate or modify the state as needed (this is going to be the
+    # part where your module will do what it needs to do)
+    protocol = None
```
protocol = 'https'

Akasurde reviewed (commented):
Few nit-pick

kbreit committed:
Rewrite environmental variable code for auth_key

auth_key code uses fallback option in the module_args statement instead of manual method.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+
+description:
+    - Enables management of Cisco's Meraki line of products through their cloud based dashboard.
+    - More information about the Meraki API can be found at U(https://dashboard.meraki.com/api_docs).
```
I would move this to the module notes.

kbreit replied:
Is that not module notes? What is the difference between that location and module notes?

dagwieers replied:
No, you also have:
```yaml
notes:
- More information about the Meraki API can be found at...
```
And these are places at the bottom of the documentation. Since I believe people looking at the module docs are interested in: 1. What does the module do 2. How does it work 3. Where's more information

PS I do think the notes should come before the return values in order of importance, so I am going to propose to change this.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+            - The HTTP method of the request.
+            - Using C(delete) is typically used for deleting objects.
+            - Using C(get) is typically used for querying objects.
+            - Using C(put) is typically used for modifying objects.
```
So I am interested to know the difference between PUT and POST. And highlight that here.

kbreit replied:
For the sake of discussion, https://stackoverflow.com/questions/107390/whats-the-difference-between-a-post-and-a-put-http-request#107450 has a nice description. There are some subtle differences, but interestingly enough, one of which is idempotency (PUT is idempotent). I'm not sure if it makes sense to explain the differences in the documentation since that's pretty in the weeds. Meraki's documentation does explain whether `POST` or `PUT` should be used for each action. When I do more abstracted modules, I'll need to work out which to use internally, largely by handling idempotency manually.

dagwieers replied:
So I know the difference between POST and PUT from an HTTP perspective. The question was how Meraki implements it. If you don't want to explain the difference, we could add:
```yaml
- Using C(post) and C(put) for modifying objects (as specified in the Meraki documentation).
```

BTW Reading the statements again, they also sound weird: `Using XXX is typically used for YYYYing objects.`. I would make that `Use XXX for YYYYing objects.`

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+        default: yes
+    content:
+        description:
+            - Raw content which should be fed in body.
```
So, in aci_rest we actually allow both JSON as well as YAML, and YAML support, especially inline, is very useful to make this module appear more pure-Ansible.

kbreit replied:
That's on my list of features to add prior to releasing the module. I just haven't gotten there quite yet.

dagwieers replied:
Perfect !

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+- name: Query network device inventory
+  sda_rest:
+    auth_key: abc12345
+    host: dashboard.meraki.com
```
I would default to api.meraki.com, so that people don't have to provide this information.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+    host: dashboard.meraki.com
+    method: get
+    path: /api/v0/organizations
+    use_ssl: yes
```
HTTPS is default, so not required.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+    # define the available arguments/parameters that a user can pass to
+    # the module
+    module_args = dict(auth_key=dict(type='str', no_log=True, fallback=(env_fallback, ['MERAKI_KEY'])),
+                       host=dict(type='str', required=True),
```
I prefer this:
```python
        host=dict(type='str', default='api.meraki.com'),
```

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+                       validate_certs=dict(type='bool', default=True),
+                       content=dict(type='raw'),
+                       src=dict(type='path'),
+                       output_level=dict(type='str', choices=['normal', 'debug'])
```
I'd go for:
```python
        output_level=dict(type='str', default='normal', choices=['normal', 'debug']),
```

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+               'X-Cisco-Meraki-API-Key': module.params['auth_key'],
+               }
+
+    if module.params['output_level'] == 'debug':
```
So, I think on failure this output would be useful as well. So maybe it's better to create a new dictionary `debug_result` which you can then provide to fail_json calls related to fetch_url failing, and you can merge it with result before exit_json.

```python
debug_result = dict(
    url=url,
    method=module.params['method'].upper(),
    headers=headers,
    payload=payload,
)
```

Then add the response and status:
```python
debug_result['status'] = info['status']
debug_result['response'] = resp.read()
```

Then call fail_json using it:
```python
    module.fail_json(msg=str(e), **debug_result)
```

And exit_json:
```python
    if module.params['output_level'] == 'debug':
        result.update(debug_result)

    module.exit_json(**result)
```

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+            result['response_headers'] = str(resp.headers)
+
+    except Exception as e:
+        module.fail_json(msg=e)
```
You need to ensure it is a string, so msg=str(e) !

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+        try:
+            result['message'] = json.loads(to_native(resp.read()))
+        except:
+            module.fail_json(msg="Meraki dashboard didn't return JSON compatible data")
```
It's not Meraki dashboard, but Meraki API that didn't return JSON compatible data.

We have to get rid of any relation to Dashboard IMO. Dashboard is the GUI.

kbreit replied:
Once again, I'm conflicted on the nomenclature I should use. Meraki has multiple APIs and this one in particular is called Dashboard API. I need to think it through a little more but Meraki Dashboard API may be the best bet. Otherwise it's ambiguous to which of Meraki's APIs it may refer to. I'll make a change though before it's merged though.

dagwieers replied:
Oh, ok.

dagwieers commented on lib/ansible/modules/network/meraki/meraki_rest.py:
```diff
@@ -0,0 +1,238 @@
+    except Exception as e:
+        module.fail_json(msg=e)
+
+    if info['status'] >= 300:
```
Why not move this down as the `else` in the below if-statement ?

sivel commented:
To give fair warning, there is a possibility that this module may not be accepted as it does not meet our recent guideline additions that state that generic API modules will not be accepted.

See http://docs.ansible.com/ansible/devel/dev_guide/developing_modules.html#should-you-develop-a-module item (5)

kbreit commented:
@sivel Thanks for the heads up. I'm not completely wed to this module and am fine moving my effort into more defined modules. Would you recommend that's what I should do instead of this module?

sivel commented:
@kbreit Creating more defined modules is definitely our recommendation, and would involve less resistance on the path to getting them accepted.

kbreit commented:
@sivel Thanks for the feedback. I'll keep this PR open for a little bit and create defined modules with new PRs. FWIW, I also have a similar DNA Center module which I will treat the same way.

dagwieers commented:
@sivel I think that recommendation is ill-advised (and it is the first I hear about it), simply because in most cases implementing the complete stack of the API is not feasible. E.g. for ACI it would mean thousands of modules if you want 100% coverage of functionality. So the idea is to provide modules that cover the majority of the common needs, leave aci_rest for the remaining needs, but allow people to contribute. So a common REST module would be the first module to implement, and the more dedicated modules then slowly cover common cases.

BTW The **aci_rest** module is idempotent and reports change.

sivel commented:
@dagwieers We're not saying that they shouldn't exist, just that we will not ship them with ansible.  We will be moving towards pushing many new modules submissions and existing modules over time to use galaxy as a distribution method.

kbreit commented:
@sivel I know we're digressing into a pretty big conversation in a PR...but...how will it be determined what goes into Ansible and what stays in Galaxy? Will Galaxy have quality assurance mechanisms? This may be me old-at-heart, but I'm not a fan of downloading components from public repositories and praying they work (ex. Docker Hub).

dagwieers commented:
@sivel Well, before this kind of decisions or advice is being expressed, we should probably wait until all pieces are lined up. I'm going to keep my opinion on Galaxy in its current form to myself, but without a cleanup and moderation, and topics like integration testing, release management, trust and security, community-building and governance, etc. There's so much complexity and issues that will reflect potentially bad on Ansible and Ansible's user experience that we're far from making this happen. Just moving stuff elsewhere because we (Red Hat + existing community) are not able to handle it in Ansible is not a solution, it's going to be the same problems, but multiplied in scope and impact.

kbreit committed:
Small changes

dagwieers committed:
Changes requested from the preview

This PR includes:
- Review documentation, some options are less useful
- Default to api.meraki.com
- Returned data in 'data' key
- Support debug output on failures
- Add 'url', 'method', 'headers', 'status', 'response' to debug output
- Report no change on queries (GET)
- Improved error handling

dagwieers committed:
More improvements

This PR includes:
- Add a list of notes
- Add an inline JSON example
- Warn when SSL is disabled
- Remove check-mode support
- Require one of src/content

dagwieers committed:
Documentation fix

dagwieers committed:
Fix parameter documentation

dagwieers committed:
Simplification

dagwieers committed:
Add inline YAML and JSON support

This PR includes:
- Inline YAML and JSON support
- Default method to 'get'
- Require yaml library for YAML file or YAML string support

dagwieers committed:
Asorted fixes to inline JSON/YAML support and output

kbreit committed:
Removed an error condition

kbreit commented:
Closing for now as this is not in development.
